### PR TITLE
Add fullscreen CLI option for SDL renderer

### PR DESCRIPTION
**Pull request by MichaelMackus**
Better way to do this would likely be to override the font scale when in fullscreen mode. This way, we can automatically change the font scale a good setting. However, the current results look great to me in 1080p with 1.5 font scale.

Related #221

### CODE REVIEW

**Comment by Mikolaj** [comment]
Oh, cool, well done.

Regarding override, I'd rather keep commandline and config commands explicit and with separation of concerns. However, it would be nice to add, in another issue, an in-game keybinding that binds to a composition of three human commands (or more?): "set internal font scale to the system's default resolution, saving the old resolution somewhere", "reload the current fontset with the current font scale, saving the old fonts and their texture atlases in a map indexed by scale; if already in the map, just restore instead of reloading" and "go fullscreen". Preferably the commands would have inverses and then the keybinding could toggle fullscreen and we could also have a toggle 'fullscreen on/off' in the settings menu that just executes the keybinding. Feel free to open any issues about that and I'd help to flesh them out.

File: engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs [hunk 1]
```diff
@@ -189,7 +189,9 @@ startupFun coscreen soptions@ClientOptions{..} rfMVar = do
   SDL.initialize [SDL.InitVideo]
   let screenV2 = SDL.V2 (toEnum $ rwidth coscreen * boxSize)
                         (toEnum $ rheight coscreen * boxSize)
-      windowConfig = SDL.defaultWindow {SDL.windowInitialSize = screenV2}
+      windowConfig = SDL.defaultWindow
+        { SDL.windowInitialSize = screenV2
+        , SDL.windowMode = if sfullscreen then SDL.Fullscreen else SDL.Windowed }
```
Comment (Mikolaj) [inline engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:194]: OOI, what would `FullscreenDesktop` do here?

File: engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs [hunk 2]
```diff
@@ -198,6 +200,7 @@ startupFun coscreen soptions@ClientOptions{..} rfMVar = do
         }
   swindow <- SDL.createWindow title windowConfig
   srenderer <- SDL.createRenderer swindow (-1) rendererConfig
+  when sfullscreen (SDL.rendererLogicalSize srenderer SDL.$= Just screenV2)
```
Comment (Mikolaj) [inline engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:203]: Why is this needed? Given that `srenderer` was created based on a window created with `screenV2` size, I'd expect it to be already set, but I generally don't understand this setting and its function, so I may be missing something. OTOH, if we change the font scale in-game I fully expect we'd need to call `rendererLogicalSize`.

File: engine-src/Game/LambdaHack/Server/Commandline.hs [hunk 3]
```diff
@@ -184,6 +186,12 @@ newGameP = optional $ max 1 <$> min difficultyBound <$>
               <> help "Start a new game, overwriting the save file, with difficulty for all UI players set to N"
               <> metavar "N" )
 
+fullscreenP :: Parser Bool
+fullscreenP =
+  switch (  long "fullscreen"
+         <> short 'f'
+         <> help "Start game in fullscreen mode (currently, only implemented for the SDL renderer)" )
```
Comment (Mikolaj) [inline engine-src/Game/LambdaHack/Server/Commandline.hs:193]: We warn in README that only SDL and Web is maintained (if it's not clear, needs to clarified), so I'd simplify and remove the disclaimer about SDL. And in in the browser, while the power-users can change the config, it's not supported, and also I don't think they'd expect fullscreen or scaling to work, when they can do that via browser fullscreen or zoom. BTW, we'd need to CPP away, for the browser build, the fullscreen toggle in Settings menu, if we implement it.

File: engine-src/Game/LambdaHack/Server/Commandline.hs [hunk 4]
```diff
@@ -86,6 +87,7 @@ serverOptionsP = do
       sclientOptions = ClientOptions
         { sfonts         = []  -- comes only from config file
         , sfontsets      = []  -- comes only from config file
+        , sfullscreen    = sfullscreenSer
```
Comment (Mikolaj) [inline engine-src/Game/LambdaHack/Server/Commandline.hs:90]: I think you can remove this line if you name it `sfullscreen` instead of `sfullscreenSer` above. I don't think it's one of the rare cases where we need a separate server and client clone of a setting.

File: engine-src/Game/LambdaHack/Common/ClientOptions.hs [hunk 5]
```diff
@@ -69,6 +71,7 @@ defClientOptions = ClientOptions
   , sallFontsScale = Nothing
   , sfonts = []
   , sfontsets = []
+  , sfullscreen = False
```
Comment (Mikolaj) [inline engine-src/Game/LambdaHack/Common/ClientOptions.hs:74]: Could you also, in a separate commit, add this to config file, just following closely how it's done for `sallFontsScale`?

**Comment by MichaelMackus** [comment]
This is needed so that the game "stretches" to take up the full screen
space +if font scale is set improperly). This also centers the game
vertically (so it isnt in the top left corner).

**Comment by MichaelMackus** [comment]
Yes, I was looking to add to config file but ran out of time.

File: engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs (hunk 1 above)
Comment (MichaelMackus) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:194]: That is  a "fake fullscreen" according to SDL wiki. So, I'm assuming, fullscreen windowed mode with no borders. We can add that as a separate config/CLI option, but I left it out here for simplicity.

File: engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs (hunk 2 above)
Comment (MichaelMackus) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:203]: This is needed so that the game "stretches" to take up the full screen
space (if font scale is set improperly, you are on retina screen, etc.). This also centers the game
vertically (so it isn't in the top left corner) and adjusts all the maths internally so we don't have to add offsets & scaling math to our UI code.

File: engine-src/Game/LambdaHack/Common/ClientOptions.hs (hunk 5 above)
Comment (MichaelMackus) [reply engine-src/Game/LambdaHack/Common/ClientOptions.hs:74]: Yes, I was looking to add to config file but ran out of time.

File: engine-src/Game/LambdaHack/Server/Commandline.hs (hunk 4 above)
Comment (MichaelMackus) [reply engine-src/Game/LambdaHack/Server/Commandline.hs:90]: Cool, I'm not very familiar with newer GHC features, but I do see the triple dots below which I believe brings in the local vars. Working on this now.

File: engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs (hunk 2 above)
Comment (Mikolaj) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:203]: Oh, neat. Let's keep it.

File: engine-src/Game/LambdaHack/Server/Commandline.hs (hunk 4 above)
Comment (Mikolaj) [reply engine-src/Game/LambdaHack/Server/Commandline.hs]: It's RecordWildCards + NamedFieldPuns. Enabled by default is the whole codebase. Error-prone, unless you are aware of it and you actively manage it. Then it's quite concise, sometime, but not always, making up for the boilerplate that lens eliminates completely (at it's own cost).

File: engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs (hunk 1 above)
Comment (Mikolaj) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:194]: I've done some searching and it seems each of the modes don't work for some people. Do you think it would make sense to rename the option to windowMode and let the user specify any of the values that SDL accepts? The type has Read instance, so it should work be fine. In config, by default, Windowed could be set.

Comment (Mikolaj) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:194]: Actually, it's easier to just define own 3-value type. Otherwise, each frontend module would need to provide a (usually erroring out) parser for the type, I think. Too complex.

Comment (MichaelMackus) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:194]: I can look into this. So you are saying you want me to add different CLI options for e.g. `--fullscreen` and `--fullscreen-windowed`?

Comment (Mikolaj) [reply engine-src/Game/LambdaHack/Client/UI/Frontend/Sdl.hs:194]: Thank you, I would be grateful. I guess I'd prefer to have the same schema in the config file and on command line. And in the config file, probably `windowMode = FullscreenDesktop` is natural and obvious. So `--windowMode FullscreenDesktop`.

**Comment by Mikolaj** [comment]
@MichaelMackus: hi! how is it going? need any help?

**Comment by Mikolaj** [comment]
@MichaelMackus: any news? If you are too busy, I will just merge this and then modify according to our discussion in separate commits.

**Comment by MichaelMackus** [comment]
Hey, sorry I have been busy with other projects and haven't had a chance to
wrap this up.

**Comment by Mikolaj** [comment]
No problem at all. It's not a duty, but fun and games. :)

What are you up to?

**Comment by MichaelMackus** [comment]
Just working on my day job and playing with some libp2p stuff in my free
time. Once I have some more time I'll take a look again, I did want to play
with some ideas around absracting rulesets and whatnot.
